internal/k8s/reference_checkers.go

```diff
@@ -262,3 +261,43 @@ func isPolicyReferenced(policies []v1.PolicyReference, resourceNamespace string,
 
 	return false
 }
+
+type dosResourceReferenceChecker struct {
+	annotation string
+}
+
+func newDosResourceReferenceChecker(annotation string) *dosResourceReferenceChecker {
+	return &dosResourceReferenceChecker{annotation}
+}
+
+func (rc *dosResourceReferenceChecker) IsReferencedByIngress(namespace string, name string, ing *networking.Ingress) bool {
+	res, exists := ing.Annotations[rc.annotation]
+	if !exists {
+		return false
+	}
+	return res == namespace+"/"+name || (namespace == ing.Namespace && res == name)
+}
+
+func (rc *dosResourceReferenceChecker) IsReferencedByMinion(_ string, _ string, _ *networking.Ingress) bool {
+	return false
+}
+
+func (rc *dosResourceReferenceChecker) IsReferencedByVirtualServer(namespace string, name string, vs *v1.VirtualServer) bool {
+	if vs.Spec.Dos == namespace+"/"+name || (namespace == vs.Namespace && vs.Spec.Dos == name) {
+		return true
+	}
+	for _, route := range vs.Spec.Routes {
+		if route.Dos == namespace+"/"+name || (namespace == vs.Namespace && route.Dos == name) {
+			return true
+		}
+	}
+	return false
+}
+
+func (rc *dosResourceReferenceChecker) IsReferencedByVirtualServerRoute(_ string, _ string, _ *v1.VirtualServerRoute) bool {
+	return false
```

Comment:
we should now support VSRs, right?

Reply:
oh yeah. Shame we don't have a test to capture this.

Reply:
fixed